src/Processors/Transforms/Streaming/JoinTransformWithAlignment.cpp

```diff
@@ -202,79 +173,138 @@ IProcessor::Status JoinTransformWithAlignment::prepare()
+            if (isCancelled())
+                return;
+
+            auto & [retracted_chunk, chunk] = left_input.input_chunks.front();
+            if (chunk.maxTimestamp() + latency_threshold < right_input.minTimestamp() || right_input_in_quiesce)
             {
-                /// right input already progressed ahead enough or it is in quiesce, it's time to join
-                for (auto & left_chunk : left_input.input_chunks)
-                {
-                    if (likely(left_chunk.hasRows()))
-                    {
-                        auto joined_block = left_input.input_port->getHeader().cloneWithColumns(left_chunk.detachColumns());
-                        join->joinLeftBlock(joined_block);
-
-                        if (auto rows = joined_block.rows(); rows > 0)
-                            output_chunks.emplace_back(joined_block.getColumns(), rows);
-                    }
-                    else
-                    {
-                        /// FIXME, watermark / heartbeat / checkpoint chunk.
-                    }
-                }
-
-                left_input.input_chunks.clear();
-
-                /// Continue reading more data from left input
-                left_input.muted = false;
+                if (retracted_chunk.rows())
+                    processInputData<true>(retracted_chunk);
+
+                if (likely(chunk.rows()))
+                    processInputData<true>(chunk.chunk);
+
+                left_input.input_chunks.pop_front();
 
                 if (right_input_in_quiesce)
-                    right_input.muted = false;
+                    unmuteInput(right_input);
             }
             else
             {
-                /// Left stream is leading ahead, mute left input to wait right input to progress
-                left_input.muted = true;
-                right_input.muted = false;
-                stats.left_input_muted += 1;
+                unmuteInput(right_input);
+                break;
             }
         }
-        else
-        {
-            /// Left input doesn't have any data yet, we like to pull some. Mute the right input to
-            /// give left input more chance to catch up.
-            left_input.muted = false;
+    } while (left_input.hasValidInputs() && right_input.hasValidInputs());
 
-            /// If the right input doesn't have watermark yet, don't mute it
-            if (likely(right_input.watermark != INVALID_WATERMARK))
-            {
-                right_input.muted = true;
-                stats.right_input_muted += 1;
-            }
-        }
+    if (!left_input.hasValidInputs())
+    {
+        /// Left input has not valid chunks, unmute it
+        unmuteInput(left_input);
+        /// If right input has at least one valid chunk (another one may be an incomplete updated chunk pair <retracted_chunk, null>),
+        /// mute right input to avoid buffer too much data.
+        /// FIXME: allow buffer more ? limited by rows or bytes.
```

Comment:
If we need alignment, we can't buffer more otherwise the alignment may never be aligned since we are always progressing the watermark if we buffer more ?

Reply:
`if (hasCompleteChunks()) muteInput(...)` may is better